util/versionutils.c

```diff
@@ -0,0 +1,333 @@
+          index_v++;
+          continue;
+        }
+
+      if (is_digit != g_ascii_isdigit (ver[index_v]))
+        {
+          is_digit = !is_digit;
+          if (index_pv > 0 && prep_version[index_pv - 1] != '.')
+            {
+              prep_version[index_pv] = '.';
+              index_pv++;
+            }
+        }
+
+      if (ver[index_v] == 'r')
+        {
+          if (strstr (ver + index_v, "releasecandidate") == ver + index_v)
+            {
+              prep_version[index_pv] = 'r';
+              prep_version[index_pv + 1] = 'c';
+              index_pv += 2;
+              index_v += 16;
+              continue;
+            }
+          if ((strstr (ver + index_v, "release-candidate") == ver + index_v)
+              || (strstr (ver + index_v, "release_candidate") == ver + index_v))
+            {
+              prep_version[index_pv] = 'r';
+              prep_version[index_pv + 1] = 'c';
+              index_pv += 2;
+              index_v += 17;
+              continue;
+            }
+        }
+
+      prep_version[index_pv] = ver[index_v];
+      index_v++;
+      index_pv++;
+    }
+
+  prep_version[index_pv] = '\0';
+  g_free (ver);
+  return (g_strdup (prep_version));
+}
+
+/**
+ * @brief Gets the release state of a specified part of the version string
+ *        if any.
+ *
+ * @param[in]  version  The version string to get the release state from.
+ * @param[in]  index    The part of the version string to check.
+ *
+ * @return  Returns 0 if there is no release state, returns 4 if the release
+ *          state is "development" (dev), returns 3 if the state is "alpha",
+ *          2 if the state is beta and 1 if the state is release candidate (rc).
+ */
+static int
+get_release_state (const char *version, int index)
+{
+  char *part;
+  int rel_stat = 0;
+
+  part = get_part (version, index);
+
+  if (part == NULL)
+    return (0);
+
+  if (strcmp (part, "dev") == 0 || strcmp (part, "development") == 0)
+    rel_stat = 4;
+  if (strcmp (part, "alpha") == 0)
+    rel_stat = 3;
+  if (strcmp (part, "beta") == 0)
+    rel_stat = 2;
+  if (strcmp (part, "rc") == 0)
+    rel_stat = 1;
+
+  g_free (part);
+  return (rel_stat);
+}
+
+/**
+ * @brief Gets the part of the version string that is specified by index.
+ *
+ * @param[in]  version  The version string to get the part from.
+ * @param[in]  index    The part of the version string to return.
+ *
+ * @return  Returns a copy of the specified part of the version string.
+ */
+static char *
+get_part (const char *version, int index)
+{
+  int dot_count = 0;
+  int begin, end;
+
+  for (begin = 0; begin < (int) strlen (version) && dot_count < index; begin++)
+    {
+      if (version[begin] == '.')
+        dot_count++;
+    }
+
+  if (begin == (int) strlen (version))
+    return (NULL);
+
+  for (end = begin + 1; end < (int) strlen (version) && version[end] != '.';
+       end++)
+    ;
+
+  return (str_cpy ((char *) (version + begin), end - begin));
```

Comment:
How about using `g_strdup`?

Reply:
There is [g_strndup](https://docs.gtk.org/glib/func.strndup.html). Thanks for the changes.